packages/dds/tree/src/changeset/toDelta.ts

```diff
@@ -0,0 +1,362 @@
+
+/**
+ * Converts inserted content into the format expected in Delta instances.
+ * This involves applying the following changes:
+ * - Updating node values
+ * - Inserting new subtrees within the inserted content
+ * - Deleting parts of the inserted content
+ *
+ * The only kind of change that is not applied by this function is MoveIn.
+ *
+ * @param node - The subtree to apply modifications to. Updated in place.
+ * @param modify - The modifications to either apply or collect.
+ * @returns The remaining modifications that the consumer of the Delta will apply on the given node. May be empty if
+ *   all modifications are applied by the function.
+ */
+function applyOrCollectModifications(
+    node: Delta.ProtoNode,
+    modify: Pick<T.Modify, "value" | "fields">,
+): InsertedFieldsMarksMap {
+    const outFieldsMarks: InsertedFieldsMarksMap = new Map();
+    if (modify.value !== undefined) {
+        const type = modify.value.type;
+        switch (type) {
+            case "Set":
+                node.value = modify.value.value;
+                break;
+            case "Revert":
+                fail(ERR_REVERT_ON_INSERT);
+            default: unreachableCase(type);
+        }
+    }
+    if (modify.fields !== undefined) {
+        const protoFields = node.fields ?? new Map();
+        const modifyFields = modify.fields;
+        for (const key of Object.keys(modifyFields)) {
+            const brandedKey = brand<FieldKey>(key);
+            const outNodes = protoFields.get(brandedKey) ?? fail(ERR_MOD_ON_MISSING_FIELD);
+            const outMarks: InsertedFieldsMarks = [];
+            let index = 0;
+            let offset = 0;
+            for (const markWithOffset of modifyFields[key]) {
+                index += markWithOffset.offset || 0;
+                offset += markWithOffset.offset || 0;
+                const mark = markWithOffset.mark;
+                if (Array.isArray(mark)) {
+                    for (const attach of mark) {
+                        // Inline into `switch(attach.type)` once we upgrade to TS 4.7
+                        const type = attach.type;
+                        switch (type) {
+                            case "Insert": {
+                                const content = cloneTreeContent(attach.content);
+                                outNodes.splice(index, 0, ...content);
+                                index += content.length;
+                                offset += content.length;
+                                break;
+                            }
+                            case "MInsert": {
+                                const cloned = cloneAndModify(attach);
+                                if (cloned.fields.size > 0) {
+                                    outMarks.push({
+                                        offset,
+                                        mark: {
+                                            type: Delta.MarkType.Modify,
```

Comment:
Why is the type Modify and not InsertAndModify?

Reply:
Because in the delta format we inline/flatten nested inserts' contents into the root insert's contents.